diff --git a/device/src/controller.c b/device/src/controller.c
index 60820c2c..2e8cf630 100644
--- a/device/src/controller.c
+++ b/device/src/controller.c
@@ -346,7 +346,7 @@ static const bt_version_t *get_bt_version(void) {
 // TODO(zachoverflow): hide inside, move decoder inside too
 static const bt_device_features_t *get_features_classic(int index) {
   assert(readable);
-  assert(index < MAX_FEATURES_CLASSIC_PAGE_COUNT);
+  //assert(index < MAX_FEATURES_CLASSIC_PAGE_COUNT);
   return &features_classic[index];
 }
 
